hail/src/main/scala/is/hail/expr/ir/EmitStream.scala

```diff
@@ -1266,41 +1358,51 @@ object EmitStream {
             SizedStream(setup, newStream, len)
           }
 
-        case StreamLeftJoinDistinct(leftIR, rightIR, leftName, rightName, compIR, joinIR) =>
-          val lEltType = coerce[PStream](leftIR.pType).elementType
-          val rEltType = coerce[PStream](rightIR.pType).elementType.setRequired(false)
-          val xLElt = mb.newEmitField("join_lelt", lEltType)
-          val xRElt = mb.newEmitField("join_relt", rEltType)
-
-          val env2 = env.bind(leftName -> xLElt, rightName -> xRElt)
+        case StreamJoinRightDistinct(leftIR, rightIR, lKey, rKey, leftName, rightName, joinIR, joinType) =>
+          assert(joinType == "left" || joinType == "outer")
+          val lEltType = coerce[PStruct](coerce[PStream](leftIR.pType).elementType)
+          val rEltType = coerce[PStruct](coerce[PStream](rightIR.pType).elementType)
+          val xLElt = mb.newEmitField("join_lelt", lEltType.orMissing(joinType == "left"))
+          val xRElt = mb.newEmitField("join_relt", rEltType.setRequired(false))
+          val newEnv = env.bind(leftName -> xLElt, rightName -> xRElt)
+
+          val lKeyViewType = PSubsetStruct(lEltType, lKey: _*)
```

Comment:
nice. I was gonna suggest this if you didn't use it, it's totally the right thing to do here!

Prefer lKey.toArray though, that's probably the better constructor.

cc @akotlar

Reply:
Yeah, I wrote something similar for the iterator joins long ago. I certainly don't want to make a copy of the key on every comparison! I was very happy to find `PSubsetStruct`!